### ISSUE TITLE
Optimize Exact_cond_midP_unspecific_ordering_rx2()

### ISSUE THREAD

**Issue by wleoncio** [ocbe-uio/contingencytables#44]
Current version makes fairly trivial examples like `the_2xc_table(fontanella_2008)` take forever to run. Probable cause: lots of doubly-nested `for`-loops.

**Comment by wleoncio** [comment]
After profiling, the issue seems to be mostly due to a call to the `calc_Pvalue_4x2.ExactCond_unspecific()` method (ditto for 5x2, which is not triggered by the `fontanella_2008` example). The aforementioned method contains a triply-nested for-loop, which R doesn't like.

**Comment by wleoncio** [comment]
For benchmarking, these examples run fast:

```R
font_div_10 <- ceiling(fontanella_2008 / 10)
Exact_cond_midP_unspecific_ordering_rx2(t(font_div_10), "increasing")
the_2xc_table(font_div_10)
```

**Comment by wleoncio** [comment]
Profiling results:

```r
$by.self
                                       self.time self.pct total.time total.pct
"as.character"                            126.86    35.10     126.86     35.10
"grepl"                                    78.68    21.77     208.82     57.77
"test_statistic"                           45.58    12.61     113.56     31.42
"apply"                                    36.64    10.14      57.80     15.99
"findInCallstack"                          11.94     3.30     221.62     61.31
"aperm"                                     6.22     1.72       8.92      2.47
"identical"                                 5.36     1.48       5.36      1.48
"convertFunName2Method"                     4.94     1.37     231.74     64.11
"calc_prob.ExactCond"                       4.56     1.26       4.80      1.33
"calc_Pvalue_4x2.ExactCond_unspecific"      4.12     1.14     361.44     99.99
"rbind"                                     3.58     0.99       4.08      1.13
"is.factor"                                 3.28     0.91       3.28      0.91
"calc_prob"                                 2.90     0.80     239.44     66.24
"FUN"                                       2.62     0.72       2.62      0.72
"as.list"                                   2.26     0.63       4.82      1.33
"aperm.default"                             2.24     0.62       2.70      0.75
"unlist"                                    2.10     0.58       2.10      0.58
"matrix"                                    1.82     0.50       1.86      0.51
"sys.calls"                                 1.78     0.49       1.78      0.49
"c"                                         1.28     0.35       1.28      0.35
"length"                                    1.26     0.35       1.26      0.35
"prod"                                      1.24     0.34       1.24      0.34
"isTRUE"                                    1.16     0.32       1.66      0.46
"match.fun"                                 1.12     0.31       1.20      0.33
"lengths"                                   1.10     0.30       1.10      0.30
"as.list.default"                           0.78     0.22       0.78      0.22
">"                                         0.68     0.19       0.68      0.19
":"                                         0.66     0.18       0.66      0.18
"any"                                       0.60     0.17       0.60      0.17
"/"                                         0.58     0.16       0.58      0.16
"-"                                         0.48     0.13       0.48      0.13
"^"                                         0.40     0.11       0.40      0.11
"class<-"                                   0.36     0.10       0.36      0.10
"is.na"                                     0.36     0.10       0.36      0.10
"+"                                         0.34     0.09       0.34      0.09
"("                                         0.26     0.07       0.26      0.07
"*"                                         0.26     0.07       0.26      0.07
"anyNA"                                     0.26     0.07       0.26      0.07
"dim"                                       0.18     0.05       0.18      0.05
"is.recursive"                              0.18     0.05       0.18      0.05
"<"                                         0.12     0.03       0.12      0.03
"=="                                        0.10     0.03       0.10      0.03
"is.function"                               0.08     0.02       0.08      0.02
"is.atomic"                                 0.04     0.01       0.04      0.01
"names"                                     0.04     0.01       0.04      0.01
"dimnames"                                  0.02     0.01       0.02      0.01
"findCenvVar"                               0.02     0.01       0.02      0.01
"min"                                       0.02     0.01       0.02      0.01

$by.total
                                          total.time total.pct self.time self.pct
"Exact_cond_midP_unspecific_ordering_rx2"     361.46    100.00      0.00     0.00
"calc_Pvalue_4x2.ExactCond_unspecific"        361.44     99.99      4.12     1.14
"calc_prob"                                   239.44     66.24      2.90     0.80
"convertFunName2Method"                       231.74     64.11      4.94     1.37
"findInCallstack"                             221.62     61.31     11.94     3.30
"grepl"                                       208.82     57.77     78.68    21.77
"as.character"                                126.86     35.10    126.86    35.10
"test_statistic"                              113.56     31.42     45.58    12.61
"apply"                                        57.80     15.99     36.64    10.14
"aperm"                                         8.92      2.47      6.22     1.72
"identical"                                     5.36      1.48      5.36     1.48
"as.list"                                       4.82      1.33      2.26     0.63
"calc_prob.ExactCond"                           4.80      1.33      4.56     1.26
"rbind"                                         4.08      1.13      3.58     0.99
"is.factor"                                     3.28      0.91      3.28     0.91
"aperm.default"                                 2.70      0.75      2.24     0.62
"FUN"                                           2.62      0.72      2.62     0.72
"unlist"                                        2.10      0.58      2.10     0.58
"matrix"                                        1.86      0.51      1.82     0.50
"sys.calls"                                     1.78      0.49      1.78     0.49
"isTRUE"                                        1.66      0.46      1.16     0.32
"c"                                             1.28      0.35      1.28     0.35
"length"                                        1.26      0.35      1.26     0.35
"prod"                                          1.24      0.34      1.24     0.34
"match.fun"                                     1.20      0.33      1.12     0.31
"lengths"                                       1.10      0.30      1.10     0.30
"as.list.default"                               0.78      0.22      0.78     0.22
">"                                             0.68      0.19      0.68     0.19
":"                                             0.66      0.18      0.66     0.18
"any"                                           0.60      0.17      0.60     0.17
"/"                                             0.58      0.16      0.58     0.16
"-"                                             0.48      0.13      0.48     0.13
"^"                                             0.40      0.11      0.40     0.11
"class<-"                                       0.36      0.10      0.36     0.10
"is.na"                                         0.36      0.10      0.36     0.10
"+"                                             0.34      0.09      0.34     0.09
"("                                             0.26      0.07      0.26     0.07
"*"                                             0.26      0.07      0.26     0.07
"anyNA"                                         0.26      0.07      0.26     0.07
"dim"                                           0.18      0.05      0.18     0.05
"is.recursive"                                  0.18      0.05      0.18     0.05
"<"                                             0.12      0.03      0.12     0.03
"=="                                            0.10      0.03      0.10     0.03
"is.function"                                   0.08      0.02      0.08     0.02
"is.atomic"                                     0.04      0.01      0.04     0.01
"names"                                         0.04      0.01      0.04     0.01
"dimnames"                                      0.02      0.01      0.02     0.01
"findCenvVar"                                   0.02      0.01      0.02     0.01
"min"                                           0.02      0.01      0.02     0.01
"cmp"                                           0.02      0.01      0.00     0.00
"cmpCall"                                       0.02      0.01      0.00     0.00
"cmpfun"                                        0.02      0.01      0.00     0.00
"compiler:::tryCmpfun"                          0.02      0.01      0.00     0.00
"constantFold"                                  0.02      0.01      0.00     0.00
"constantFoldCall"                              0.02      0.01      0.00     0.00
"doTryCatch"                                    0.02      0.01      0.00     0.00
"genCode"                                       0.02      0.01      0.00     0.00
"getFoldFun"                                    0.02      0.01      0.00     0.00
"getInlineInfo"                                 0.02      0.01      0.00     0.00
"h"                                             0.02      0.01      0.00     0.00
"isBaseVar"                                     0.02      0.01      0.00     0.00
"tryCatch"                                      0.02      0.01      0.00     0.00
"tryCatchList"                                  0.02      0.01      0.00     0.00
"tryCatchOne"                                   0.02      0.01      0.00     0.00
"tryInline"                                     0.02      0.01      0.00     0.00

$sample.interval
[1] 0.02

$sampling.time
[1] 361.46
```

**Comment by wleoncio** [comment]
Slowest functions (see below) point to problem being in the `findInCallstack()` function, which will be addressed in issue #49. in the meantime, methods should be called directly.

```
$by.self
                                       self.time self.pct total.time total.pct
"as.character"                            126.86    35.10     126.86     35.10
"grepl"                                    78.68    21.77     208.82     57.77
"test_statistic"                           45.58    12.61     113.56     31.42
"apply"                                    36.64    10.14      57.80     15.99
"findInCallstack"                          11.94     3.30     221.62     61.31
```

**Comment by wleoncio** [comment]
Indeed, `$sampling.time` was reduced to 53.1 after calling the method directly. Other optimizations were performed to smaller effect.
